core/chains/evm/gas/models.go

```diff
@@ -261,32 +264,53 @@ func (e *evmFeeEstimator) L1Oracle() rollups.L1Oracle {
 	return e.EvmEstimator.L1Oracle()
 }
 
-func (e *evmFeeEstimator) GetFee(ctx context.Context, calldata []byte, feeLimit uint64, maxFeePrice *assets.Wei, opts ...feetypes.Opt) (fee EvmFee, chainSpecificFeeLimit uint64, err error) {
+func (e *evmFeeEstimator) GetFee(ctx context.Context, calldata []byte, feeLimit uint64, maxFeePrice *assets.Wei, toAddress *common.Address, opts ...feetypes.Opt) (fee EvmFee, chainSpecificFeeLimit uint64, err error) {
+	// Create call msg for gas limit estimation, if needed
+	callMsg := ethereum.CallMsg{
+		To:   toAddress,
+		Data: calldata,
+	}
+
 	// get dynamic fee
 	if e.EIP1559Enabled {
 		var dynamicFee DynamicFee
 		dynamicFee, err = e.EvmEstimator.GetDynamicFee(ctx, maxFeePrice)
 		if err != nil {
 			return
 		}
-		chainSpecificFeeLimit, err = commonfee.ApplyMultiplier(feeLimit, e.geCfg.LimitMultiplier())
 		fee.DynamicFeeCap = dynamicFee.FeeCap
 		fee.DynamicTipCap = dynamicFee.TipCap
-		return
+		chainSpecificFeeLimit = feeLimit
+		// Set call msg fields with dynamic fee fields for gas limit estimation, if needed
+		callMsg.GasFeeCap = fee.DynamicFeeCap.ToInt()
+		callMsg.GasTipCap = fee.DynamicTipCap.ToInt()
+	} else {
+		// get legacy fee
+		fee.Legacy, chainSpecificFeeLimit, err = e.EvmEstimator.GetLegacyGas(ctx, calldata, feeLimit, maxFeePrice, opts...)
+		if err != nil {
+			return
+		}
+		// Set call msg fields with legacy fee field for gas limit estimation, if needed
+		callMsg.GasPrice = fee.Legacy.ToInt()
 	}
 
-	// get legacy fee
-	fee.Legacy, chainSpecificFeeLimit, err = e.EvmEstimator.GetLegacyGas(ctx, calldata, feeLimit, maxFeePrice, opts...)
-	if err != nil {
-		return
+	if e.geCfg.EstimateGasLimit() {
+		callMsg.Gas = chainSpecificFeeLimit
```

Comment:
1. We don't have to set a gas limit to get an estimation. This can be ommited.
2. I'm interested in knowing what is the exact requirement from the product teams regarding gas limit estimation. Personally, I would remove any requirements for GasLimit limits if `EstimateGasLimit` is enabled. Usually, when we have fallback values people tend to set them wrong because they're not aware what is the impact. I've observed the same pattern with the gas estimators where we have default values that are only used once.

Reply:
1. I was using it as a way to cap the estimation to the limit products sent us but I'll remove this and add that restriction in code for better flexibility
2. At least for the initial solution that's needed for Mantle, I'm trying to honor the gas limits products provide as a cap to avoid any changes in behavior/assumptions. Functions has mentioned that they would need us to honor the cap otherwise it could lead to failures at the contract level. We can discuss more in the thread though.